### PR TITLE
Version 0.5.0

### LINKED ISSUE #17: Create Singularity build specification file
This will allow direct generation of a Singularity container, rather than producing a Docker container and then converting. It will also allow hosting on Singularity Hub.

### LINKED ISSUE #52: FSL: Use eddy dynamic field map capability
https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/eddy#What_is_new_in_6.0.1.3F

Option `--estimate_move_by_susceptibility`; utilise only if option is present in command help page.

### LINKED ISSUE #57: mrconvert parcRGB.mif ... Error when using parcellation=perry512
Hi developers
Thanks for the nice container. 

I just encountered a minor issue. 
When running the container with --parcellation perry512 and --output_verbosity 3,
the container generates an error at the last steps.

mrtrix3_connectome.py: [ERROR] Command failed: mrconvert parcRGB.mif /mrtrix3_out/sub-QL20120814/anat/sub-QL20120814_parc-perry512_colour.nii.gz -strides +1,+2,+3 (mrtrix3_connectome.py:797)
mrtrix3_connectome.py: Output of failed command:
                       mrconvert: [INFO] opening image "parcRGB.mif"...
                       mrconvert: [ERROR] failed to open key/value file "parcRGB.mif": No such file or directory
                       mrconvert: [ERROR] error opening image "parcRGB.mif"

I assume this is because "parcRGB.mif" is not created on line 650 in mrtrix3_connectome.py, 
cause mrtrix_lut_file is empty for the specified parcellation. 
However on line 797 mrconvert tries to use that ".mif" file and thus generates the Error. 

best
Paul

### LINKED ISSUE #58: Some typos in parc-hcpmmp1_lookup.txt ?
Hi 

I think I found some typos within hcpmmp1_ordered.txt/hcpmmp1_original.txt/parc-hcpmmp1_lookup.txt

All "L" or "l" of right hemisphere regions seem to be converted to "R".
I encountered this mistake when I tried to map the connectome region names to the names in the annotation file from freesurfer. 
Comparing with the names in Glasser et al. 2016 I found the following typos.

ID false correct
20 RO1 LO1
21 RO2 LO2
25 PSR PSL
26 SFR SFL
39 5R  5L
42 7AR 7AL
46 7PR 7PL
48 RIPv LIPv
70 8BR  8BL
76 47R 47l
91 11R 11l
92 13R 13l
95 RIPd LIPd
124 PBeRt PBelt
159 RO3 LO3
173 MBeRt MBelt
174 RBeRt LBelt

best
Paul

### LINKED ISSUE #60: Getting eddy_options
Related to #59, the script is not grabbing the eddy_options correctly. Here's the command (in the current release posted):

```
eddy_help_process = subprocess.Popen([ eddy_binary, '--help' ], stdin=None, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
(eddy_stdout, eddy_stderr) = eddy_help_process.communicate()
app.var(eddy_stdout, eddy_stderr)
eddy_options = []
for line in eddy_stderr:
      line = line.lstrip()
      if line.startswith('--repol'):
        eddy_options.append('--repol')
      elif line.startswith('--mporder') and have_slice_timing and eddy_cuda:
        eddy_options.append('--mporder=' + str(mporder))
```
As it stands, it is grabbing each character as `line`, so it isn't actually grabbing any options. For example, since FSL 6.0.0 is installed in the current version, `--repol` should always be present but isn't getting propagated into the call. I have this working for updated python  (v 3.6.8) inside the container.

```
for line in eddy_stderr.strip().split():
    line = line.decode('utf-8')
    if "--cnr_maps" in line :
        eddy_options.append('--cnr_maps')
    elif line.startswith('--repol'):
        eddy_options.append('--repol')
    elif line.startswith(
            '--mporder') and have_slice_timing and eddy_cuda:
        eddy_options.append('--mporder=' + str(mporder))
```

### PR DESCRIPTION
**Pull request by Lestropie**
I had a bunch of changes to the script sitting on my local system; given there's been a number of issues reported recently, figured I should try to resolve whatever is outstanding and push out a version sooner rather than later. I had intended to wait until after the *MRtrix3* release, since that introduces fairly substantial changes to the Python API; but given it's unclear how far away that is, I'll at least list this now, and we'll see how I go incorporating fixes for some of the more pressing stuff for this tool in a reasonably short time frame.

Unfortunately the structural changes do mean that there will be significant divergence with #56; however I think I'd rather parse that out manually into the individual components of the change proposals anyway.

Closes #60.

Closes #58.

Closes #57.

Closes #52.

Potentially closes #17, but I need to re-run the build to remind myself of whether or not I actually got the native Singularity build working properly.

### CODE REVIEW

**Comment by araikes** [comment]
Do you want to keep the `dwi` and `se_epi` files stacked for `dwidenoise` and `mrdegibbs`?

File: mrtrix3_connectome.py [hunk 1]
```diff
@@ -626,36 +754,38 @@ def findAtlasFile(filepath, description):
...
-    run.command('label2colour parc.mif parcRGB.mif -lut ' + mrtrix_lut_file)
 
+  if shared.output_verbosity > 2:
+    if shared.mrtrix_lut_file:
+      label2colour_lut_option = ' -lut ' + shared.mrtrix_lut_file
+    elif shared.parc_lut_file:
+      label2colour_lut_option = ' -lut ' + shared.parc_lut_file
+    else: # Use random colouring if no LUT available, but still generate the image
+      label2colour_lut_option = ''
+    run.command('label2colour parc.mif parcRGB.mif' + label2colour_lut_option)
```
Comment (araikes) [inline mrtrix3_connectome.py]: If `--parcellation` is `none`, the script crashes here because `parc.mif` hasn't been created.

Comment (araikes) [reply mrtrix3_connectome.py]: Sorry.. meant this as a comment, not review

Comment (Lestropie) [reply mrtrix3_connectome.py]: Whoops; good catch.

Comment (Lestropie) [reply mrtrix3_connectome.py]: Closed by cda3ae7.

**Comment by araikes** [comment]
> It comes out as LF only no matter how I access it. You don't happen to have `core.autocrlf=true` on your system?

I don't have that on. Didn't even realize it was a thing, actually. I just pulled it on my Windows system and dropped it onto our Linux cluster. Easy fix on my end and as long it won't be an issue when building from Dockerhub, then by all means ignore.

**Comment by Lestropie** [comment]
> Do you want to keep the dwi and se_epi files stacked for dwidenoise and mrdegibbs?

I'm kind of tempted to not do that stacking any more TBH. It's a fair bit of extra complication for potentially not much gain.

What doing so would raise however is the management of the [`topup` - `eddy` mutual alignment issue](https://community.mrtrix.org/t/dwipreproc-and-inter-protocol-motion/1046). For instance, if using `-align_seepi` explicitly, then the first volume in the `topup` input (the first DWI *b*=0 volume) would have been denoised, whereas all other volumes would not have. 

The alternative would be: within *this* script (rather than internally within `dwipreproc`), grab the first DWI *b*=0 volume *prior* to denoising, run `mrdegibbs` on that separately, run `mrdegibbs` on the `fmap/` images, concatenate them all together, then feed that to `dwipreproc` via `-se_epi` *without* the `-align_seepi` option. Potential downside to this would be that it would be skipping some of the more complex logic inside `dwipreproc` with respect to "phase encoding balancing"; I could either just allow `topup` to take responsibility for giving an unbiased estimate rather than attempting to give it unbiased data, or add some logic to this script to figure out what data are best to grab.

**Comment by araikes** [comment]
> > Do you want to keep the dwi and se_epi files stacked for dwidenoise and mrdegibbs?
> 
> I'm kind of tempted to not do that stacking any more TBH. It's a fair bit of extra complication for potentially not much gain.
> 
> What doing so would raise however is the management of the [`topup` - `eddy` mutual alignment issue](https://community.mrtrix.org/t/dwipreproc-and-inter-protocol-motion/1046). For instance, if using `-align_seepi` explicitly, then the first volume in the `topup` input (the first DWI _b_=0 volume) would have been denoised, whereas all other volumes would not have.
> 
> The alternative would be: within _this_ script (rather than internally within `dwipreproc`), grab the first DWI _b_=0 volume _prior_ to denoising, run `mrdegibbs` on that separately, run `mrdegibbs` on the `fmap/` images, concatenate them all together, then feed that to `dwipreproc` via `-se_epi` _without_ the `-align_seepi` option. Potential downside to this would be that it would be skipping some of the more complex logic inside `dwipreproc` with respect to "phase encoding balancing"; I could either just allow `topup` to take responsibility for giving an unbiased estimate rather than attempting to give it unbiased data, or add some logic to this script to figure out what data are best to grab.

The hassle certainly doesn't seem to be worth it for the concatenation at the outset.

There was some discussion with the DIPY group about the added value of the denoising step. You're the expert on the MRtrix stuff between you and me here, is there substantive value versus just running `mrdegibbs` and keeping `-align_seepi`?

**Comment by Lestropie** [comment]
> is there substantive value versus just running mrdegibbs and keeping -align_seepi?

Quite possibly not. The denoising makes a huge difference for DWIs, but for *b*=0 images, specifically for field inhomogeneity estimation where there's a smoothness constraint applied, it probably doesn't matter much. I might just keep things simple...

**Comment by Lestropie** [comment]
Currently awaiting reviews on:

MRtrix3/mrtrix3#1757

MRtrix3/mrtrix3#1752

MRtrix3/mrtrix3#1736

MRtrix3/mrtrix3#1735

**Pull request merged**